pkg/storegateway/indexheader/thread.go

```diff
@@ -0,0 +1,104 @@
+// Provenance-includes-license: MIT
+// Provenance-includes-copyright: Christian Grabowski
+
+package indexheader
+
+import (
+	"runtime"
+)
+
+type execResult struct {
+	value interface{}
+	err   error
+	// populated is only set to `true` by the start method of OSThread. The field only
+	// exists so that we can tell when we've gotten a zero value of it. This is used by
+	// the Call method so that we can detect when the thread is shutting down.
+	populated bool
+}
+
+type OSThread struct {
+	// call is used to submit closures for this thread to execute.
+	call chan func() (interface{}, error)
+
+	// res is used to send the results of running closures back to callers.
+	res chan execResult
+
+	// stopping is a request for this thread to stop running.
+	stopping chan struct{}
+
+	// stopped is confirmation that the thread has stopped running.
+	stopped chan struct{}
+}
+
+func NewOSThread() *OSThread {
+	return &OSThread{
+		call:     make(chan func() (interface{}, error), 1),
+		res:      make(chan execResult, 1),
+		stopping: make(chan struct{}),
+		stopped:  make(chan struct{}),
+	}
+}
+
+func (o *OSThread) start() {
+	runtime.LockOSThread()
+
+	defer func() {
+		runtime.UnlockOSThread()
+		close(o.stopped)
+	}()
+
+	for {
+		select {
+		case <-o.stopping:
+			// Close the channel used to read results so that any callers waiting on
+			// results (which can happen if they submitted something before this method
+			// has a chance to execute it but after this thread pool has been stopped)
+			// so that they immediately get a zero value which they can check for.
+			close(o.res)
+			return
+		case fn := <-o.call:
+			val, err := fn()
+			o.res <- execResult{value: val, err: err, populated: true}
+			break
+		}
+	}
+}
+
+func (o *OSThread) Start() {
+	go o.start()
+}
+
+// Stop indicates to the thread that it should stop accepting new work but, it
+// does *not* guarantee that all work is stopped after the method returns. If
+// you need to wait for all work to complete, call Join after Stop.
+func (o *OSThread) Stop() {
+	close(o.stopping)
+}
+
+// Join waits for the thread to complete any tasks it was executing after Stop
+// is called. This allows callers to ensure the thread has completed executing
+// which is useful in tests to verify no goroutines leak.
+func (o *OSThread) Join() {
+	<-o.stopped
+}
+
+func (o *OSThread) Call(fn func() (interface{}, error)) (interface{}, error) {
+	select {
+	case <-o.stopping:
+		// Make sure the pool (and hence this thread) hasn't been stopped before
+		// submitting our function to run. It's still possible for the thread pool
+		// to be stopped between this check and our function actually being run by
+		// the .start() method. We handle this by closing the o.res channel as part
+		// of stopping this thread and testing for a zero value being returned
+		// (which is what reading from a closed channel does).
+		return nil, ErrPoolStopped
+	case o.call <- fn:
+		res := <-o.res
```

Comment:
We can't actually use unbuffered channels:
> Actually, I don't think we can remove the buffer form o.call. If we do, the write to it in Call() will block until it's read which may not happen if the thread is in the process of stopping (i.e. .start() is no longer reading from it because o.stopping has been closed).

Reply:
> A thread can only have tasks submitted to it from a single caller at a time, after it's been acquired via the .Call() method in ThreadPool.

If there are two threads racing, then one thread can push a call, second thread can push another call  (because channel is buffered, and the first one was already read by the thread), and then we're in a situations where both callers are racing to read a result from `o.res`, in which case it can be any of them.

```mermaid
sequenceDiagram
    participant caller1
    participant caller2
    participant o.call
    participant o.res

    caller1 ->> o.call: pushes call1
    o.call ->> thread: call1
    o.res -->> caller1: waiting for result
    caller2 ->> o.call: pushes call2
    o.res -->> caller2: waiting for result
    thread ->> o.res: res1
    o.res ->> caller2: reads res1
    o.call ->> thread: call2
    thread ->> o.res: res2
    o.res  ->> caller1: reads res2

```

Reply:
This can't actually happen because functions are only pushed to `OSThread.Call` after the thread has been removed from the pool here: https://github.com/grafana/mimir/pull/1660/files#diff-74baca6d3e8d3598a308a4ec324a9e416a24a5b01c8bbd78dce0f10bdab3db49R109

Before anything is submitted to the `OSThread`, it's removed from `ThreadPool.pool` such that only a single caller of `ThreadPool.Call` is operating on it.

Reply:
Can we then unexport `OSThread` to be `osThread` since it requires `ThreadPool` logic to work safely?

Reply:
Yes, good idea